File: packages/agent/src/run-wrapper.ts
import { execFile } from 'node:child_process'
import { promisify } from 'node:util'

import type { RunStatus, RunTimelineEventPayload, RunTool, RunTurnOptions } from '@webmux/shared'
import { createClaudeClient, type ClaudeClient } from './claude-client.js'
import { ClaudeMessageParser } from './claude-event.js'
import { createCodexClient, type CodexClient } from './codex-client.js'
import { prepareCodexInput } from './codex-input.js'
import { parseCodexThreadEvent } from './codex-event.js'
import type { RunImageAttachmentUpload } from '@webmux/shared'

const execFileAsync = promisify(execFile)

export interface RunWrapperOptions {
  runId: string
  tool: RunTool
  toolThreadId?: string
  repoPath: string
  prompt: string
  attachments?: RunImageAttachmentUpload[]
  options?: RunTurnOptions
  onEvent: (status: RunStatus, summary?: string, hasDiff?: boolean) => void
  onFinish: (status: RunStatus) => void
  onItem: (item: RunTimelineEventPayload) => void
  onThreadReady?: (toolThreadId: string) => void
  codexClientFactory?: () => CodexClient
  claudeClientFactory?: () => ClaudeClient
}

export class RunWrapper {
  private readonly tool: RunTool
  private readonly toolThreadId?: string
  private readonly repoPath: string
  private readonly prompt: string
  private readonly attachments: RunImageAttachmentUpload[]
  private readonly onEvent: RunWrapperOptions['onEvent']
  private readonly onFinish: RunWrapperOptions['onFinish']
  private readonly onItem: RunWrapperOptions['onItem']
  private readonly turnOptions: RunTurnOptions
  private readonly onThreadReady?: RunWrapperOptions['onThreadReady']
  private readonly codexClientFactory: () => CodexClient
  private readonly claudeClientFactory: () => ClaudeClient
  private abortController: AbortController | null = null
  private claudeQuery: ReturnType<ClaudeClient['query']> | null = null
  private currentStatus: RunStatus = 'starting'
  private disposed = false
  private interrupted = false
  private latestSummary: string | undefined
  private finished = false

  constructor(options: RunWrapperOptions) {
    this.tool = options.tool
    this.toolThreadId = options.toolThreadId
    this.repoPath = options.repoPath
    this.prompt = options.prompt
    this.attachments = options.attachments ?? []
    this.turnOptions = options.options ?? {}
    this.onEvent = options.onEvent
    this.onFinish = options.onFinish
    this.onItem = options.onItem
    this.onThreadReady = options.onThreadReady
    this.codexClientFactory = options.codexClientFactory ?? createCodexClient
    this.claudeClientFactory = options.claudeClientFactory ?? createClaudeClient
  }

  async start(): Promise<void> {
    if (this.disposed) {
      return
    }

    if (this.tool === 'codex') {
      await this.startCodexThread()
      return
    }

    await this.startClaudeThread()
  }

  interrupt(): void {
    if ((this.tool === 'codex' && !this.abortController) || (this.tool !== 'codex' && !this.claudeQuery) || this.disposed || this.finished) {
      return
    }

    this.interrupted = true
    this.onItem({
      type: 'activity',
      status: 'warning',
      label: 'Interrupt requested',
    })
    this.emitStatus('interrupted', this.latestSummary)
    if (this.abortController) {
      this.abortController.abort()
      return
    }

    void this.claudeQuery?.interrupt().catch((error) => {
      if (this.disposed || this.finished) {
        return
      }

      const detail = error instanceof Error ? error.message : String(error)
      this.onItem({
        type: 'activity',
        status: 'warning',
        label: 'Interrupt request failed',
        detail,
      })
    })
  }

  dispose(): void {
    if (this.disposed) {
      return
    }

    this.disposed = true

    if (this.abortController) {
      this.abortController.abort()
      this.abortController = null
    }

    if (this.claudeQuery) {
      this.claudeQuery.close()
      this.claudeQuery = null
    }
  }

  private async complete(finalStatus: RunStatus): Promise<void> {
    const hasDiff = await detectRepoChanges(this.repoPath)

    if (finalStatus === 'success') {
      this.onItem({
        type: 'activity',
        status: 'success',
        label: 'Run completed',
      })
    } else if (finalStatus === 'failed') {
      this.onItem({
        type: 'activity',
        status: 'error',
        label: 'Run failed',
      })
    }

    this.finalize(finalStatus, hasDiff, this.latestSummary)
  }

  private finalize(
    finalStatus: RunStatus,
    hasDiff: boolean,
    summary?: string,
  ): void {
    if (this.finished) {
      return
    }

    this.finished = true
    this.abortController = null
    this.claudeQuery = null
    this.emitStatus(finalStatus, summary, hasDiff)
    this.onFinish(finalStatus)
  }

  private emitStatus(status: RunStatus, summary?: string, hasDiff?: boolean): void {
    this.currentStatus = status
    this.onEvent(status, summary, hasDiff)
  }

  private async startCodexThread(): Promise<void> {
    this.emitStatus('starting')
    this.onItem({
      type: 'activity',
      status: 'info',
      label: 'Starting Codex',
      detail: this.repoPath,
    })

    const client = this.codexClientFactory()
    const threadOptions = {
      workingDirectory: this.repoPath,
      skipGitRepoCheck: true,
      sandboxMode: 'workspace-write' as const,
      approvalPolicy: 'never' as const,
      networkAccessEnabled: true,
      ...(this.turnOptions.model ? { model: this.turnOptions.model } : {}),
      ...(this.turnOptions.codexEffort ? { modelReasoningEffort: this.turnOptions.codexEffort } : {}),
    }
    const resumeId = this.turnOptions.clearSession ? undefined : this.toolThreadId
    const thread = resumeId
      ? client.resumeThread(resumeId, threadOptions)
      : client.startThread(threadOptions)
    const preparedInput = await prepareCodexInput(this.prompt, this.attachments)

    this.abortController = new AbortController()
    this.emitStatus('running')

    let sawTurnCompleted = false
    let announcedThreadId = this.toolThreadId ?? undefined

    try {
      const streamedTurn = await thread.runStreamed(preparedInput.input, {
        signal: this.abortController.signal,
      })

      for await (const event of streamedTurn.events) {
        if (this.disposed || this.finished) {
          return
        }

        const result = parseCodexThreadEvent(event)
        const nextThreadId = result.threadId ?? thread.id ?? undefined
        if (nextThreadId && nextThreadId !== announcedThreadId) {
          announcedThreadId = nextThreadId
          this.onThreadReady?.(nextThreadId)
        }

        if (result.summary) {
          this.latestSummary = result.summary
          this.emitStatus(this.currentStatus, result.summary)
        }

        for (const item of result.items) {
          this.onItem(item)
        }

        if (event.type === 'turn.completed') {
          sawTurnCompleted = true
        }

        if (result.finalStatus) {
          await this.complete(result.finalStatus)
          return
        }
      }

      if (this.finished) {
        return
      }

      const finalStatus = this.interrupted
        ? 'interrupted'
        : sawTurnCompleted
          ? 'success'
          : 'failed'
      await this.complete(finalStatus)
    } catch (error) {
      if (this.disposed || this.finished) {
        return
      }

      if (this.interrupted || isAbortError(error)) {
        await this.complete('interrupted')
        return
      }

      const message = error instanceof Error ? error.message : String(error)
      this.onItem({
        type: 'activity',
        status: 'error',
        label: 'Run failed',
        detail: message,
      })
      this.finalize('failed', false, message)
    } finally {
      await preparedInput.cleanup()
    }
  }

  private async startClaudeThread(): Promise<void> {
    if (this.attachments.length > 0) {
      throw new Error('Image attachments are currently supported for Codex only')
    }

    this.emitStatus('starting')
    this.onItem({
      type: 'activity',
      status: 'info',
      label: 'Starting Claude',
      detail: this.repoPath,
    })

    const client = this.claudeClientFactory()
    const parser = new ClaudeMessageParser()
    const query = client.query(this.prompt, {
      cwd: this.repoPath,
      resume: this.turnOptions.clearSession ? undefined : this.toolThreadId,
      persistSession: true,
      permissionMode: 'bypassPermissions',
      allowDangerouslySkipPermissions: true,
      ...(this.turnOptions.model ? { model: this.turnOptions.model } : {}),
      ...(this.turnOptions.claudeEffort ? { effort: this.turnOptions.claudeEffort } : {}),
    })

    this.claudeQuery = query
    this.emitStatus('running')

    let announcedThreadId = this.toolThreadId ?? undefined

    try {
      for await (const message of query) {
        if (this.disposed || this.finished) {
          return
        }

        const nextThreadId = message.session_id
        if (nextThreadId && nextThreadId !== announcedThreadId) {
          announcedThreadId = nextThreadId
          this.onThreadReady?.(nextThreadId)
        }

        const result = parser.parse(message)
        if (result.summary) {
          this.latestSummary = result.summary
          this.emitStatus(this.currentStatus, result.summary)
        }

        for (const item of result.items) {
          this.onItem(item)
        }

        if (result.finalStatus) {
          await this.complete(this.interrupted ? 'interrupted' : result.finalStatus)
          return
        }
      }

      if (this.finished) {
        return
      }

      await this.complete(this.interrupted ? 'interrupted' : 'failed')
    } catch (error) {
      if (this.disposed || this.finished) {
        return
      }

      if (this.interrupted || isAbortError(error)) {
        await this.complete('interrupted')
        return
      }

      const message = error instanceof Error ? error.message : String(error)
      this.onItem({
        type: 'activity',
        status: 'error',
        label: 'Run failed',
        detail: message,
      })
      this.finalize('failed', false, message)
    }
  }
}

function isAbortError(error: unknown): boolean {
  if (!error || typeof error !== 'object') {
    return false
  }

  const candidate = error as { name?: string; code?: string }
  return candidate.name === 'AbortError' || candidate.code === 'ABORT_ERR'
}

async function detectRepoChanges(repoPath: string): Promise<boolean> {
  try {
    const { stdout } = await execFileAsync(
      'git',
      ['status', '--porcelain'],
      { cwd: repoPath },
    )
    return stdout.trim().length > 0
  } catch {
    return false
  }
}

File: rust/crates/webmux-agent/src/run_wrapper.rs
use std::process::Stdio;

use tokio::process::Command;
use tracing::error;

use webmux_shared::{
    RunImageAttachmentUpload, RunStatus, RunTimelineEventPayload, RunTimelineEventStatus,
    RunTool, RunTurnOptions,
};

use crate::claude::client::{
    start_claude, ClaudeClientOptions, ClaudeHandle, ClaudeStreamEvent,
};
use crate::claude::event::ClaudeMessageParser;
use crate::codex::client::{start_codex, CodexClientOptions, CodexHandle, CodexStreamEvent};
use crate::codex::event::{parse_codex_event, CodexEvent};
use crate::codex::input::prepare_codex_input;

// ---------------------------------------------------------------------------
// RunWrapper — orchestrates a single LLM execution (Claude or Codex)
// ---------------------------------------------------------------------------

pub struct RunWrapperOptions {
    #[allow(dead_code)]
    pub run_id: String,
    pub tool: RunTool,
    pub tool_thread_id: Option<String>,
    pub repo_path: String,
    pub prompt: String,
    pub attachments: Vec<RunImageAttachmentUpload>,
    pub options: RunTurnOptions,
}

/// Callbacks invoked during a run.
pub struct RunCallbacks<FEvent, FFinish, FItem, FThread>
where
    FEvent: FnMut(RunStatus, Option<String>, Option<bool>),
    FFinish: FnMut(RunStatus),
    FItem: FnMut(RunTimelineEventPayload),
    FThread: FnMut(String),
{
    pub on_event: FEvent,
    pub on_finish: FFinish,
    pub on_item: FItem,
    pub on_thread_ready: FThread,
}

/// The mutable state of a run, shared between the wrapper and the async
/// execution loop.
#[allow(dead_code)]
pub struct RunState {
    pub current_status: RunStatus,
    pub latest_summary: Option<String>,
    pub interrupted: bool,
    pub finished: bool,
    pub disposed: bool,
}

impl RunState {
    #[allow(dead_code)]
    pub fn new() -> Self {
        Self {
            current_status: RunStatus::Starting,
            latest_summary: None,
            interrupted: false,
            finished: false,
            disposed: false,
        }
    }
}

/// Active run handle. Holds the subprocess handle so it can be interrupted or
/// killed.
#[allow(dead_code)]
pub enum ActiveHandle {
    Claude(ClaudeHandle),
    Codex(CodexHandle),
    None,
}

impl ActiveHandle {
    /// Interrupt the running process.
    pub fn interrupt(&mut self) {
        match self {
            ActiveHandle::Claude(h) => h.interrupt(),
            ActiveHandle::Codex(h) => h.abort(),
            ActiveHandle::None => {}
        }
    }

    /// Kill / dispose the running process.
    pub fn dispose(&mut self) {
        match self {
            ActiveHandle::Claude(h) => h.close(),
            ActiveHandle::Codex(h) => h.abort(),
            ActiveHandle::None => {}
        }
    }
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Start a run. This spawns the appropriate subprocess and reads events until
/// completion or interruption. Callbacks are invoked inline.
///
/// Returns the active handle and run state. The caller can use the handle to
/// interrupt or dispose the run.
pub async fn start_run<FEvent, FFinish, FItem, FThread>(
    opts: RunWrapperOptions,
    callbacks: &mut RunCallbacks<FEvent, FFinish, FItem, FThread>,
) where
    FEvent: FnMut(RunStatus, Option<String>, Option<bool>),
    FFinish: FnMut(RunStatus),
    FItem: FnMut(RunTimelineEventPayload),
    FThread: FnMut(String),
{
    match opts.tool {
        RunTool::Codex => run_codex(opts, callbacks).await,
        RunTool::Claude => run_claude(opts, callbacks).await,
    }
}

// ---------------------------------------------------------------------------
// Codex run
// ---------------------------------------------------------------------------

async fn run_codex<FEvent, FFinish, FItem, FThread>(
    opts: RunWrapperOptions,
    cb: &mut RunCallbacks<FEvent, FFinish, FItem, FThread>,
) where
    FEvent: FnMut(RunStatus, Option<String>, Option<bool>),
    FFinish: FnMut(RunStatus),
    FItem: FnMut(RunTimelineEventPayload),
    FThread: FnMut(String),
{
    (cb.on_event)(RunStatus::Starting, None, None);
    (cb.on_item)(RunTimelineEventPayload::Activity {
        status: RunTimelineEventStatus::Info,
        label: "Starting Codex".to_string(),
        detail: Some(opts.repo_path.clone()),
    });

    let clear_session = opts.options.clear_session.unwrap_or(false);
    let resume_id = if clear_session {
        None
    } else {
        opts.tool_thread_id.clone()
    };

    let codex_opts = CodexClientOptions {
        working_directory: opts.repo_path.clone(),
        model: opts.options.model.clone(),
        reasoning_effort: opts.options.codex_effort.as_ref().map(|e| format!("{e:?}").to_lowercase()),
        resume_thread_id: resume_id,
    };

    // Prepare input (write images to temp files)
    let prepared = match prepare_codex_input(&opts.prompt, &opts.attachments).await {
        Ok(p) => p,
        Err(e) => {
            (cb.on_item)(RunTimelineEventPayload::Activity {
                status: RunTimelineEventStatus::Error,
                label: "Run failed".to_string(),
                detail: Some(e.clone()),
            });
            let has_diff = detect_repo_changes(&opts.repo_path).await;
            (cb.on_event)(RunStatus::Failed, Some(e), Some(has_diff));
            (cb.on_finish)(RunStatus::Failed);
            return;
        }
    };

    let mut handle = match start_codex(&prepared.prompt, &prepared.image_paths, &codex_opts) {
        Ok(h) => h,
        Err(e) => {
            prepared.cleanup().await;
            (cb.on_item)(RunTimelineEventPayload::Activity {
                status: RunTimelineEventStatus::Error,
                label: "Run failed".to_string(),
                detail: Some(e.clone()),
            });
            let has_diff = detect_repo_changes(&opts.repo_path).await;
            (cb.on_event)(RunStatus::Failed, Some(e), Some(has_diff));
            (cb.on_finish)(RunStatus::Failed);
            return;
        }
    };

    (cb.on_event)(RunStatus::Running, None, None);

    let current_status = RunStatus::Running;
    let mut latest_summary: Option<String> = None;
    let interrupted = false;
    let mut saw_turn_completed = false;
    let mut announced_thread_id = opts.tool_thread_id.clone();
    let mut final_result: Option<RunStatus> = None;

    while let Some(stream_event) = handle.rx.recv().await {
        match stream_event {
            CodexStreamEvent::Event(event) => {
                let result = parse_codex_event(&event);

                // Track thread id
                if let Some(tid) = &result.thread_id {
                    if announced_thread_id.as_ref() != Some(tid) {
                        announced_thread_id = Some(tid.clone());
                        (cb.on_thread_ready)(tid.clone());
                    }
                }

                if let Some(summary) = &result.summary {
                    latest_summary = Some(summary.clone());
                    (cb.on_event)(current_status.clone(), Some(summary.clone()), None);
                }

                for item in result.items {
                    (cb.on_item)(item);
                }

                if matches!(event, CodexEvent::TurnCompleted { .. }) {
                    saw_turn_completed = true;
                }

                if let Some(fs) = result.final_status {
                    final_result = Some(fs);
                    break;
                }
            }
            CodexStreamEvent::Done => break,
            CodexStreamEvent::Error(e) => {
                error!("Codex stream error: {e}");
                (cb.on_item)(RunTimelineEventPayload::Activity {
                    status: RunTimelineEventStatus::Error,
                    label: "Run failed".to_string(),
                    detail: Some(e.clone()),
                });
                final_result = Some(RunStatus::Failed);
                break;
            }
        }
    }

    prepared.cleanup().await;

    let final_status = final_result.unwrap_or_else(|| {
        if interrupted {
            RunStatus::Interrupted
        } else if saw_turn_completed {
            RunStatus::Success
        } else {
            RunStatus::Failed
        }
    });

    complete_run(&opts.repo_path, final_status.clone(), &latest_summary, cb).await;
}

// ---------------------------------------------------------------------------
// Claude run
// ---------------------------------------------------------------------------

async fn run_claude<FEvent, FFinish, FItem, FThread>(
    opts: RunWrapperOptions,
    cb: &mut RunCallbacks<FEvent, FFinish, FItem, FThread>,
) where
    FEvent: FnMut(RunStatus, Option<String>, Option<bool>),
    FFinish: FnMut(RunStatus),
    FItem: FnMut(RunTimelineEventPayload),
    FThread: FnMut(String),
{
    (cb.on_event)(RunStatus::Starting, None, None);
    (cb.on_item)(RunTimelineEventPayload::Activity {
        status: RunTimelineEventStatus::Info,
        label: "Starting Claude".to_string(),
        detail: Some(opts.repo_path.clone()),
    });

    let clear_session = opts.options.clear_session.unwrap_or(false);
    let resume = if clear_session {
        None
    } else {
        opts.tool_thread_id.clone()
    };

    let claude_opts = ClaudeClientOptions {
        cwd: opts.repo_path.clone(),
        resume,
        model: opts.options.model.clone(),
        effort: opts.options.claude_effort.as_ref().map(|e| format!("{e:?}").to_lowercase()),
        attachments: opts.attachments.clone(),
    };

    let mut handle = match start_claude(&opts.prompt, &claude_opts) {
        Ok(h) => h,
        Err(e) => {
            (cb.on_item)(RunTimelineEventPayload::Activity {
                status: RunTimelineEventStatus::Error,
                label: "Run failed".to_string(),
                detail: Some(e.clone()),
            });
            let has_diff = detect_repo_changes(&opts.repo_path).await;
            (cb.on_event)(RunStatus::Failed, Some(e), Some(has_diff));
            (cb.on_finish)(RunStatus::Failed);
            return;
        }
    };

    let mut parser = ClaudeMessageParser::new();
    (cb.on_event)(RunStatus::Running, None, None);

    let current_status = RunStatus::Running;
    let mut latest_summary: Option<String> = None;
    let interrupted = false;
    let mut announced_thread_id = opts.tool_thread_id.clone();
    let mut final_result: Option<RunStatus> = None;

    while let Some(stream_event) = handle.rx.recv().await {
        match stream_event {
            ClaudeStreamEvent::Message(msg) => {
                // Track session_id
                if let Some(sid) = &msg.session_id {
                    if announced_thread_id.as_ref() != Some(sid) {
                        announced_thread_id = Some(sid.clone());
                        (cb.on_thread_ready)(sid.clone());
                    }
                }

                let result = parser.parse(&msg);

                if let Some(summary) = &result.summary {
                    latest_summary = Some(summary.clone());
                    (cb.on_event)(current_status.clone(), Some(summary.clone()), None);
                }

                for item in result.items {
                    (cb.on_item)(item);
                }

                if let Some(fs) = result.final_status {
                    let status = if interrupted {
                        RunStatus::Interrupted
                    } else {
                        fs
                    };
                    final_result = Some(status);
                    break;
                }
            }
            ClaudeStreamEvent::Done => break,
            ClaudeStreamEvent::Error(e) => {
                error!("Claude stream error: {e}");
                (cb.on_item)(RunTimelineEventPayload::Activity {
                    status: RunTimelineEventStatus::Error,
                    label: "Run failed".to_string(),
                    detail: Some(e.clone()),
                });
                final_result = Some(RunStatus::Failed);
                break;
            }
        }
    }

    let final_status = final_result.unwrap_or_else(|| {
        if interrupted {
            RunStatus::Interrupted
        } else {
            RunStatus::Failed
        }
    });

    complete_run(&opts.repo_path, final_status, &latest_summary, cb).await;
}

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

async fn complete_run<FEvent, FFinish, FItem, FThread>(
    repo_path: &str,
    final_status: RunStatus,
    latest_summary: &Option<String>,
    cb: &mut RunCallbacks<FEvent, FFinish, FItem, FThread>,
) where
    FEvent: FnMut(RunStatus, Option<String>, Option<bool>),
    FFinish: FnMut(RunStatus),
    FItem: FnMut(RunTimelineEventPayload),
    FThread: FnMut(String),
{
    let has_diff = detect_repo_changes(repo_path).await;

    match &final_status {
        RunStatus::Success => {
            (cb.on_item)(RunTimelineEventPayload::Activity {
                status: RunTimelineEventStatus::Success,
                label: "Run completed".to_string(),
                detail: None,
            });
        }
        RunStatus::Failed => {
            (cb.on_item)(RunTimelineEventPayload::Activity {
                status: RunTimelineEventStatus::Error,
                label: "Run failed".to_string(),
                detail: None,
            });
        }
        _ => {}
    }

    (cb.on_event)(
        final_status.clone(),
        latest_summary.clone(),
        Some(has_diff),
    );
    (cb.on_finish)(final_status);
}

/// Detect uncommitted changes in a git repo via `git status --porcelain`.
pub async fn detect_repo_changes(repo_path: &str) -> bool {
    let output = Command::new("git")
        .args(["status", "--porcelain"])
        .current_dir(repo_path)
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .output()
        .await;

    match output {
        Ok(o) => !o.stdout.is_empty(),
        Err(_) => false,
    }
}
